File: route/view_acl.py
from .tool.func import *

def view_acl(name):
    with get_db_connect() as conn:
        curs = conn.cursor()

        check_ok = ''
        user_page = 0
        ip = ip_check()

        if flask.request.method == 'POST':
            check_data = 'document_set (' + name + ')'
        else:
            check_data = None

        user_data = re.search(r'^user:(.+)$', name)
        if user_data:
            if check_data and ip_or_user(ip) != 0:
                return redirect('/login')

            if user_data.group(1) != ip:
                if admin_check(5) != 1:
                    if check_data:
                        return re_error('/error/3')
                    else:
                        check_ok = 'disabled'
            else:
                user_page = 1
        else:
            if admin_check(5) != 1:
                if check_data:
                    return re_error('/error/3')
                else:
                    check_ok = 'disabled'

        if flask.request.method == 'POST':
            acl_data = ['decu', 'document_edit_acl', 'document_move_acl', 'document_delete_acl', 'dis', 'view', 'why']

            for i in acl_data:
                form_data = flask.request.form.get(i, '')
            
                curs.execute(db_change("delete from acl where title = ? and type = ?"), [name, i])
                curs.execute(db_change("insert into acl (title, data, type) values (?, ?, ?)"), [name, form_data, i])
                
                curs.execute(db_change("delete from data_set where doc_name = ? and doc_rev = ? and set_name = 'acl_date'"), [name, i])
                    
                time_limit = flask.request.form.get(i + '_date', '')
                if re.search(r'^[0-9]{4}-[0-9]{2}-[0-9]{2}$', time_limit):
                    curs.execute(db_change("insert into data_set (doc_name, doc_rev, set_name, set_data) values (?, ?, 'acl_date', ?)"), [name, i, time_limit])

            all_d = ''
            for i in acl_data[:-1]:
                if flask.request.form.get(i, '') == '':
                    all_d += 'normal'
                    if i != 'view':
                        all_d += ' | '
                else:
                    all_d += flask.request.form.get(i, '')
                    if i != 'view':
                        all_d += ' | '

            markup_data = flask.request.form.get('document_markup', '')

            curs.execute(db_change("select set_data from data_set where doc_name = ? and set_name = 'document_markup'"), [name])
            db_data = curs.fetchall()

            curs.execute(db_change("delete from data_set where doc_name = ? and set_name = 'document_markup'"), [name])
            curs.execute(db_change("insert into data_set (doc_name, doc_rev, set_name, set_data) values (?, '', 'document_markup', ?)"), [name, markup_data])

            if not db_data or db_data[0][0] != markup_data:
                curs.execute(db_change("select data from data where title = ?"), [name])
                db_data_2 = curs.fetchall()
                if db_data_2:
                    render_set(
                        doc_name = name,
                        doc_data = db_data_2[0][0],
                        data_type = 'backlink'
                    )

            markup_data = markup_data if markup_data != '' else 'normal'

            if user_page == 1:
                admin_check(5, check_data + ' (' + all_d + ')' + ' (' + markup_data + ')')

            conn.commit()

            return redirect('/acl/' + url_pas(name))
        else:
            data = '<h2>' + load_lang('acl') + '</h2>'
            acl_list = get_acl_list('user') if re.search(r'^user:', name) else get_acl_list()
            if not re.search(r'^user:', name):
                acl_get_list = [
                    [load_lang('view_acl'), 'view', '3'],
                    [load_lang('document_acl'), 'decu', '4'],
                    [load_lang('document_edit_acl'), 'document_edit_acl', '5'],
                    [load_lang('document_move_acl'), 'document_move_acl', '5'],
                    [load_lang('document_delete_acl'), 'document_delete_acl', '5'],
                    [load_lang('discussion_acl'), 'dis', '3'],
                ]
            else:
                acl_get_list = [
                    [load_lang('document_acl'), 'decu', '2']
                ]

            for i in acl_get_list:
                data += '' + \
                    '<h' + i[2] + '>' + i[0] + '</h' + i[2] + '>' + \
                    '<hr class="main_hr">' + \
                    '<select name="' + i[1] + '" ' + check_ok + '>' + \
                ''

                curs.execute(db_change("select data from acl where title = ? and type = ?"), [name, i[1]])
                acl_data = curs.fetchall()
                for data_list in acl_list:
                    check = 'selected="selected"' if acl_data and acl_data[0][0] == data_list else ''
                    data += '<option value="' + data_list + '" ' + check + '>' + (data_list if data_list != '' else 'normal') + '</option>'

                data += '</select>'
                data += '<hr class="main_hr">'
                
                date_value = ''
                
                curs.execute(db_change("select set_data from data_set where doc_name = ? and doc_rev = ? and set_name = 'acl_date'"), [name, i[1]])
                db_data = curs.fetchall()
                if db_data:
                    date_value = db_data[0][0]
                
                data += '<input type="date" ' + check_ok + ' value="' + date_value + '" name="' + i[1] + '_date" pattern="\\d{4}-\\d{2}-\\d{2}">'
                data += '<hr class="main_hr">'

            curs.execute(db_change("select data from acl where title = ? and type = ?"), [name, 'why'])
            acl_data = curs.fetchall()
            acl_why = html.escape(acl_data[0][0]) if acl_data else ''
            data += '' + \
                '<h3>' + load_lang('why') + '</h3>' + \
                '<input value="' + acl_why + '" ' + check_ok + ' placeholder="' + load_lang('why') + '" name="why" ' + check_ok + '>' + \
                '<hr class="main_hr">' + \
            ''

            data += '''
                <h3>''' + load_lang('explanation') + '''</h3>
                <span id="exp"></span>
                <ul class="opennamu_ul">
                    <li>normal : ''' + load_lang('unset') + '''</li>
                    <li>admin : ''' + load_lang('admin_acl') + '''</li>
                    <li>user : ''' + load_lang('member_acl') + '''</li>
                    <li>50_edit : ''' + load_lang('50_edit_acl') + '''</li>
                    <li>all : ''' + load_lang('all_acl') + '''</li>
                    <li>email : ''' + load_lang('email_acl') + '''</li>
                    <li>owner : ''' + load_lang('owner_acl') + '''</li>
                    <li>ban : ''' + load_lang('ban_acl') + '''</li>
                    <li>before : ''' + load_lang('before_acl') + '''</li>
                    <li>30_day : ''' + load_lang('30_day_acl') + '''</li>
                    <li>ban_admin : ''' + load_lang('ban_admin_acl') + '''</li>
                    <li>not_all : ''' + load_lang('not_all_acl') + '''</li>
                    <li>90_day : ''' + load_lang('90_day_acl') + '''</li>
                </ul>
                <hr class="main_hr">
                <h2>''' + load_lang('markup') + '''</h2>
            '''

            curs.execute(db_change("select set_data from data_set where doc_name = ? and set_name = 'document_markup'"), [name])
            db_data = curs.fetchall()
            markup_load = db_data[0][0] if db_data and db_data[0][0] != '' else ''

            markup_list = ['normal'] + get_init_set_list('markup')['list']
            markup_html = ''
            for for_a in markup_list:
                if markup_load == for_a:
                    markup_html = '<option value="' + (for_a if for_a != 'normal' else '') + '">' + for_a + '</option>' + markup_html
                else:
                    markup_html += '<option value="' + (for_a if for_a != 'normal' else '') + '">' + for_a + '</option>'
            
            markup_html = '<select name="document_markup" ' + check_ok + '>' + markup_html + '</select>'

            data += markup_html
            data += '<hr class="main_hr">'

            return easy_minify(flask.render_template(skin_check(),
                imp = [name, wiki_set(), wiki_custom(), wiki_css(['(' + load_lang('acl') + ')', 0])],
                data = '''
                    <form method="post">
                        <a href="/setting/acl">(''' + load_lang('main_acl_setting') + ''')</a>
                        <hr class="main_hr">
                        ''' + render_simple_set(data) + '''
                        <button type="submit" ''' + check_ok + '''>''' + load_lang('save') + '''</button>
                    </form>
                ''',
                menu = [
                    ['w/' + url_pas(name), load_lang('return')], 
                    ['manager', load_lang('admin')], 
                    ['list/admin/auth_use_page/1/' + url_pas('acl (' + name + ')'), load_lang('acl_record')]
                ]
            ))

File: route/view_acl.go
package route

import (
	"database/sql"
	"net/url"
	"strings"

	"opennamu/route/tool"
)

var document_acl_fields = []string{
	"view",
	"decu",
	"document_edit_acl",
	"document_move_acl",
	"document_delete_acl",
	"dis",
}

var document_acl_group_fields = []string{
	"view",
	"decu",
	"document_edit_acl",
	"document_move_acl",
	"document_delete_acl",
	"dis",
}

func acl_field_title(db *sql.DB, field string) string {
	key := map[string]string{
		"view":                "view_acl",
		"decu":                "document_acl",
		"document_edit_acl":   "document_edit_acl",
		"document_move_acl":   "document_move_acl",
		"document_delete_acl": "document_delete_acl",
		"dis":                 "discussion_acl",
	}[field]
	return tool.Get_language(db, key, true)
}

func acl_value(db *sql.DB, doc_name string, field string) string {
	value := tool.Get_acl_data_list(db, doc_name, field)
	if len(value) == 0 {
		return ""
	}
	return value[0]
}

func acl_group_select(db *sql.DB) string {
	data := `<select name="acl_group">`
	for _, group := range acl_value_list(db, "") {
		data += `<option value="` + tool.HTML_escape(group) + `">` + tool.HTML_escape(group) + `</option>`
	}
	return data + `</select>`
}

func acl_group_setting(db *sql.DB, title string, field string) string {
	data := `<h3>` + acl_field_title(db, field) + `</h3>`
	groups := tool.Get_acl_data_list(db, title, field)
	if len(groups) == 0 {
		data += `<div>` + tool.Get_language(db, "normal", true) + `</div>`
	}
	for _, group := range groups {
		data += `<div>` + tool.HTML_escape(group) + ` <form method="post" style="display:inline"><input type="hidden" name="name" value="` + tool.HTML_escape(title) + `"><input type="hidden" name="acl_action" value="delete"><input type="hidden" name="acl_field" value="` + tool.HTML_escape(field) + `"><input type="hidden" name="acl_group" value="` + tool.HTML_escape(group) + `"><button type="submit">` + tool.Get_language(db, "delete", true) + `</button></form></div>`
	}
	data += `<form method="post"><input type="hidden" name="name" value="` + tool.HTML_escape(title) + `"><input type="hidden" name="acl_action" value="add"><input type="hidden" name="acl_field" value="` + tool.HTML_escape(field) + `">` + acl_group_select(db) + ` <button type="submit">` + tool.Get_language(db, "add", true) + `</button></form>`
	return data + `<hr class="main_hr">`
}

func acl_group_multiple_setting(db *sql.DB) string {
	data := `<h3>` + tool.Get_language(db, "acl", true) + `</h3><form method="post"><textarea class="opennamu_textarea_500" name="title_name" placeholder="` + tool.Get_language(db, "many_delete_help", true) + `"></textarea><hr class="main_hr"><select name="acl_field">`
	for _, field := range document_acl_group_fields {
		data += `<option value="` + tool.HTML_escape(field) + `">` + tool.HTML_escape(acl_field_title(db, field)) + `</option>`
	}
	data += `</select> ` + acl_group_select(db) + ` <button name="acl_action" value="add" type="submit">` + tool.Get_language(db, "add", true) + `</button> <button name="acl_action" value="delete" type="submit">` + tool.Get_language(db, "delete", true) + `</button></form>`
	return data
}

func View_acl(config tool.Config, doc_name string, multiple bool, values url.Values) string {
	db := tool.DB_connect()
	defer tool.DB_close(db)

	if values != nil && doc_name == "" && !multiple {
		doc_name = values.Get("name")
	}
	if doc_name == "" && !multiple && values == nil {
		data := `<form method="post"><input name="name" placeholder="` + tool.Get_language(db, "document_name", true) + `"><hr class="main_hr"><button type="submit">` + tool.Get_language(db, "go", true) + `</button></form>`
		return tool.Get_template(db, config, tool.Get_language(db, "document_setting", true), data, []any{}, [][]any{{"manager", tool.Get_language(db, "return", true)}}, map[string]string{})
	}

	if values == nil {
		allowed := tool.Check_permission(db, "document_acl_manage", config.IP)
		if strings.HasPrefix(doc_name, "user:") && strings.TrimPrefix(doc_name, "user:") == config.IP {
			allowed = true
		}
		if !allowed {
			return tool.Get_error_page(db, config, "auth")
		}
	}

	if values != nil {
		api_data := Api_acl_post(config, doc_name, multiple, values)
		response, _ := api_data["response"].(string)
		if response == "require auth" {
			return tool.Get_error_page(db, config, "auth")
		}
		if response == "redirect" {
			return tool.Get_redirect("/acl")
		}
		if response != "ok" {
			error_name, _ := api_data["data"].(string)
			if error_name == "" {
				error_name = "error"
			}
			return tool.Get_error_page(db, config, error_name)
		}
		if multiple {
			return tool.Get_redirect("/manager")
		}
		return tool.Get_redirect("/acl/" + tool.Url_parser(doc_name))
	}

	data := `<form method="post">`
	if multiple {
		data += `<textarea class="opennamu_textarea_500" name="title_name" placeholder="` + tool.Get_language(db, "many_delete_help", true) + `"></textarea><hr class="main_hr">`
	} else {
		data += `<input type="hidden" name="name" value="` + tool.HTML_escape(doc_name) + `">`
	}

	for _, field := range document_acl_fields {
		date_value := ""
		if !multiple {
			date_value = tool.Get_document_setting_value(db, doc_name, "acl_date", field)
		}
		data += `<input type="date" name="` + field + `_date" value="` + tool.HTML_escape(date_value) + `"><hr class="main_hr">`
	}

	why := ""
	if !multiple {
		why = acl_value(db, doc_name, "why")
	}
	data += `<h3>` + tool.Get_language(db, "why", true) + `</h3><input name="why" value="` + tool.HTML_escape(why) + `"><hr class="main_hr">`

	markup := ""
	if !multiple {
		markup = tool.Get_document_markup(db, doc_name, "document")
	}
	data += `<h2>` + tool.Get_language(db, "markup", true) + `</h2>` + tool.Get_markup_select_ui(db, config, doc_name, markup, "", "")
	top_disabled := ""
	if tool.Check_permission(db, "owner", config.IP) {
		top_disabled = ` disabled`
	}
	data += `<h2>` + tool.Get_language(db, "document_top", true) + `</h2><textarea class="opennamu_textarea_100" name="document_top"` + top_disabled + `>` + tool.HTML_escape(document_set_value(db, doc_name, "document_top")) + `</textarea>`
	data += `<h2>` + tool.Get_language(db, "document_editor_top", true) + `</h2><textarea class="opennamu_textarea_100" name="document_editor_top"` + top_disabled + `>` + tool.HTML_escape(document_set_value(db, doc_name, "document_editor_top")) + `</textarea><hr class="main_hr">`
	data += `<button type="submit">` + tool.Get_language(db, "save", true) + `</button></form>`
	if multiple {
		data += acl_group_multiple_setting(db)
	} else {
		for _, field := range document_acl_group_fields {
			data += acl_group_setting(db, doc_name, field)
		}
	}

	menu := [][]any{{"manager", tool.Get_language(db, "admin", true)}}
	if !multiple {
		menu = [][]any{{"w/" + tool.Url_parser(doc_name), tool.Get_language(db, "return", true)}, {"acl_multiple", tool.Get_language(db, "mutiple_document_setting", true)}, {"manager", tool.Get_language(db, "admin", true)}}
	}
	title := doc_name
	if multiple {
		title = tool.Get_language(db, "mutiple_document_setting", true)
	}
	return tool.Get_template(db, config, title, data, []any{}, menu, map[string]string{})
}

func document_set_value(db *sql.DB, doc_name string, set_name string) string {
	return tool.Get_document_setting_value(db, doc_name, set_name, "")
}
